File: Data Structures & Algorithms/validate-parentheses/submission-2.py
class Solution:
    def isValid(self, s: str) -> bool:
        st = []
        for x in s:
            if x in '([{':
                st.append(x)
            elif st:
                if x == ')' and st[-1] == '(':
                    st.pop()
                elif x == '}' and st[-1] == '{':
                    st.pop()
                elif x == ']' and st[-1] == '[':
                    st.pop()
                else:
                    return False
            else:
                return False
        if not st:
            return True
        return False

File: Data Structures & Algorithms/trapping-rain-water/submission-2.java
class Solution {
    public int trap(int[] height) {
        int ans = 0;

        Stack<Integer> st = new Stack<>();
        for (int i = 0; i < height.length; i++) {
            while (!st.isEmpty() && height[st.peek()] < height[i]) {
                int ind = height[st.pop()];
                if (!st.isEmpty()) {
                    int left = st.isEmpty() ? 0 : height[st.peek()];
                    int right = height[i];
                    int h = Math.min(left, right) - ind;
                    int w = i - st.peek() - 1;
                    ans += (h * w);
                }
            }
            st.push(i);
        }
        return ans;
    }
}
